src/polynomial.h

```diff
@@ -506,12 +487,34 @@ void Polynomial<T>::mod(Polynomial<T>* b)
     _div(nullptr, this, &a, b);
 }
 
+/** Derivate polynomial
+ * \f{eqnarray*}{
+ * P(X) &= a_0 + a_1 X + a_2 X^2 + ... + a_n X^n \\
+ * P'(X)&=       a_1   + a_2 X   + ... + a_n X^{n-1}
+ * \f}
+ *
+ * Note, derivate of \f$X^n\f$ is defined as sum of \f$n\f$ polynomials
+ * \f$X^{n-1}\f$, not a multiplication of \f$n\f$ and \f$X^{n-1}\f$. It can be
+ * expressed as below:
+ * \f{eqnarray*}{
+ *  derivative(X^n) &= (1 + 1 + ... + 1) X^{n-1}
+ *                  &= (n % p) * X^{n-1}
+ * \f{eqnarray*}{
+ * where \f$p\f$ is the characteristic of the field.
+ *
+ * @return
+ */
 template <typename T>
 void Polynomial<T>::derivative()
 {
-    Polynomial<T> a(field);
-    a.copy(this);
-    _derivative(this, &a);
+    auto it = terms.begin();
+    if (it->first == 0)
+        it++;
+    for (; it != terms.end(); it++) {
```

Comment:
use `++it` instead of `it++`: this can avoid an useless copy of the iterator (if the compiler doesn't optimize it already).